crates/engine/src/ext.rs

```diff
@@ -358,12 +404,38 @@ impl Engine {
         unimplemented!("off-chain environment does not yet support `call`");
     }
 
-    pub fn weight_to_fee(&self, _gas: u64, _output: &mut &mut [u8]) {
-        unimplemented!("off-chain environment does not yet support `weight_to_fee`");
+    /// Emulates gas price calculation.
+    pub fn weight_to_fee(&self, gas: u64, output: &mut &mut [u8]) {
+        let fee = self
+            .chain_spec
+            .gas_price
+            .saturating_mul(gas.try_into().unwrap_or(u128::MAX));
+        let fee: Vec<u8> = scale::Encode::encode(&fee);
+        set_output(output, &fee[..])
     }
 
-    pub fn random(&self, _subject: &[u8], _output: &mut &mut [u8]) {
-        unimplemented!("off-chain environment does not yet support `random`");
+    /// Returns a randomized hash.
+    ///
+    /// # Note
+    ///
+    /// - This is the off-chain environment implementation of `random`.
+    ///   It provides the same behavior in that it will likely yield the
+    ///   same hash for the same subjects within the same block (or
+    ///   execution context).
+    ///
+    /// - Returned hashes on the surface might appear random, however for
+    ///   testing purposes the actual implementation is quite simple and
+    ///   computes those "random" hashes by wrapping XOR of the internal
+    ///   entry hash with the eventually repeated sequence of the subject
+    ///   buffer.
+    pub fn random(&self, subject: &[u8], output: &mut &mut [u8]) {
+        let mut entropy_bytes = self.exec_context.entropy;
+        let len_entropy = entropy_bytes.len();
+        for (n, subject) in subject.iter().enumerate() {
+            let id = n % len_entropy;
+            entropy_bytes[id] = entropy_bytes[id] ^ subject ^ (n as u8);
+        }
```

Comment:
No need to do this manually. Just seed a rng:
```suggestion
        let seed = (self.exec_context.entropy, subject).encode();
        let rng = rand::rngs::StdRng::from_seed(&seed);
        let mut rng_bytes: [u8; 32] = Default::default();
        rng.fill(&rng_bytes);
```